File: ObjectiveCpp/Shader.m
//
//  Shader.m
//  OpenGLEditor
//
//  Created by Filip Kunc on 12/26/09.
//  For license see LICENSE.TXT
//

#import "Shader.h"

// modified from http://www.lighthouse3d.com/opengl/glsl/index.php?oglinfo

void ShaderLog(GLuint shader)
{
	int infologLength = 0;
	int charsWritten = 0;
	GLchar *infoLog;
	
	glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &infologLength);
	
	if (infologLength > 0)
	{
		infoLog = (GLchar *)malloc(infologLength);
		glGetShaderInfoLog(shader, infologLength, &charsWritten, infoLog);
		NSLog(@"shader log: %s",infoLog);
		free(infoLog);
	}
}

@implementation Shader

@synthesize shader, type;

+ (NSString *)fileExtensionForShaderType:(GLenum)aShaderType
{
	switch (aShaderType)
	{
		case GL_VERTEX_SHADER:
			return @"vs";
		case GL_FRAGMENT_SHADER:
			return @"fs";
		case GL_GEOMETRY_SHADER_EXT:
			return @"gs";
		default:
			return nil;
	}
}

- (id)initWithShaderType:(GLenum)aShaderType
		resourceInBundle:(NSString *)resourceInBundle
{
	self = [super init];
	if (self)
	{
		type = aShaderType;
		NSBundle *bundle = [NSBundle mainBundle];
		NSString *extension = [Shader fileExtensionForShaderType:aShaderType];
		NSString *path = [bundle pathForResource:resourceInBundle ofType:extension];
		
		NSString *contents;
		contents = [NSString stringWithContentsOfFile:path
											 encoding:NSUTF8StringEncoding
												error:NULL];
		
		const GLchar *shaderSource = [contents cStringUsingEncoding:NSASCIIStringEncoding];
		
		shader = glCreateShader(type);
		glShaderSource(shader, 1, &shaderSource, NULL);
		glCompileShader(shader);
		ShaderLog(shader);
	}
	return self;	
}

- (void)dealloc
{
	glDeleteShader(shader);
	[super dealloc];
}

@end

File: Classes/Shader.cpp
//
//  Shader.mm
//  OpenGLEditor
//
//  Created by Filip Kunc on 12/26/09.
//  For license see LICENSE.TXT
//

#include "Shader.h"
#include <stdio.h>

Shader::Shader(GLenum shaderType, const char *fileName)
{
    type = shaderType;
    
    NSBundle *bundle = [NSBundle mainBundle];
    NSString *file = [NSString stringWithUTF8String:fileName];
    NSString *extension = [file pathExtension];
    NSString *resourceInBundle = [file stringByDeletingPathExtension];
    NSString *path = [bundle pathForResource:resourceInBundle ofType:extension];
    
    NSString *contents;
    contents = [NSString stringWithContentsOfFile:path
                                         encoding:NSUTF8StringEncoding
                                            error:NULL];

    const GLchar *shaderSource = [contents cStringUsingEncoding:NSASCIIStringEncoding];

    shader = glCreateShader(type);
    glShaderSource(shader, 1, &shaderSource, NULL);
    glCompileShader(shader);
    
    GLint logLength, status;
    
    glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &logLength);
    if (logLength > 0)
    {
        GLchar *log = (GLchar*)malloc(logLength);
        glGetShaderInfoLog(shader, logLength, &logLength, log);
		printf("%s compile log:\n%s\n", fileName, log);
        free(log);
    }
    
    glGetShaderiv(shader, GL_COMPILE_STATUS, &status);
    if (status == 0)
    {
		printf("Failed to compile %s shader:\n%s\n", fileName, shaderSource);
    }
}

Shader::~Shader()
{
    glDeleteShader(shader);
}
